### PR TITLE
GH-89812: Add `pathlib._PurePathExt`

### PR DESCRIPTION
**Pull request by barneygale**
Move `__fspath__()`, `__bytes__()` and `as_uri()` methods from `PurePath` to a new `_PurePathExt` subclass. This new subclass is inherited by `PurePosixPath`, `PureWindowsPath` and `Path`.

Because `PurePath` isn't directly instantiatable (you get a `PurePosixPath` or `PureWindowsPath` instance back), this shouldn't change user-facing behaviour.

The methods must not be inherited future `tarfile.TarPath` and `pathlib.AbstractPath` classes, which will subclass `PurePath`.

- Discussion: https://discuss.python.org/t/make-pathlib-extensible/3428/94

* Issue: gh-89812

### CODE REVIEW

**Comment by barneygale** [comment]
The test naming is pretty janky. #104829 will help.

File: Lib/pathlib.py [hunk 1]
```diff
@@ -707,6 +665,57 @@ def match(self, path_pattern, *, case_sensitive=None):
...
+        if cls is PurePath:
+            cls = PureWindowsPath if os.name == 'nt' else PurePosixPath
+        return object.__new__(cls)
+
+    def __fspath__(self):
+        return str(self)
+
+    def __bytes__(self):
+        """Return the bytes representation of the path.  This is only
+        recommended to use under Unix."""
```
Comment (merwok) [inline Lib/pathlib.py:739]: Magic method docstrings are not easily accessible, could the class docstring have this mention instead?

File: Lib/pathlib.py [hunk 2]
```diff
@@ -707,6 +665,57 @@ def match(self, path_pattern, *, case_sensitive=None):
...
+
+    def __new__(cls, *args, **kwargs):
+        """Construct a PurePath from one or several strings and or existing
+        PurePath objects.  The strings and path objects are combined so as
+        to yield a canonicalized path, which is incorporated into the
+        new PurePath object.
+        """
+        if cls is PurePath:
+            cls = PureWindowsPath if os.name == 'nt' else PurePosixPath
+        return object.__new__(cls)
```
Comment (merwok) [inline Lib/pathlib.py]: Varargs and varkwargs are accepted but not used?
(I see it’s not new but moved code but am still asking)

File: Lib/pathlib.py [hunk 3]
```diff
@@ -707,6 +665,57 @@ def match(self, path_pattern, *, case_sensitive=None):
...
+            path = self.as_posix()
+        else:
+            # It's a posix path => 'file:///etc/hosts'
+            prefix = 'file://'
+            path = str(self)
+        return prefix + urlquote_from_bytes(os.fsencode(path))
+
+
 # Can't subclass os.PathLike from PurePath and keep the constructor
 # optimizations in PurePath.__slots__.
```
Comment (merwok) [inline Lib/pathlib.py]: This comment is slightly strange: how can «constructor optimizations» be found in \_\_slots\_\_?

File: Lib/pathlib.py (hunk 2 above)
Comment (barneygale) [reply Lib/pathlib.py]: `PurePath.__init__()` handles `*args`.

`Path.__init__()` additionally accepts `**kwargs` for historical reasons. If given, they're ignored and a deprecation warning is emitted.

File: Lib/pathlib.py (hunk 3 above)
Comment (barneygale) [reply Lib/pathlib.py]: This comment always puzzled me - it originally referred to a `_parse_args()` method, but that method never interfered with inheriting from `os.PathLike`.

I think the issue is that `os.PathLike` doesn't use `__slots__`, and so all subclasses have an instance dict. I may be wrong.

File: Lib/pathlib.py [hunk 4]
```diff
@@ -233,14 +233,10 @@ def __repr__(self):
...
-    """Base class for manipulating paths without I/O.
+class _BasePurePath(object):
+    """Base class for manipulating paths using only lexical operations.
 
-    PurePath represents a filesystem path and offers operations which
-    don't imply any actual filesystem I/O.  Depending on your system,
-    instantiating a PurePath will return either a PurePosixPath or a
-    PureWindowsPath object.  You can also instantiate either of these classes
-    directly, regardless of your system.
+    This class does not provide __fspath__(), __bytes__() or as_uri().
```
Comment (merwok) [inline Lib/pathlib.py]:
```suggestion
    This class does not provide the methods __fspath__, __bytes__ or as_uri.
```

File: Lib/pathlib.py [hunk 5]
```diff
@@ -707,6 +665,62 @@ def match(self, path_pattern, *, case_sensitive=None):
...
+class PurePath(_BasePurePath):
+    """Base class for manipulating paths without I/O.
+
+    PurePath represents a filesystem path and offers operations which
+    don't imply any actual filesystem I/O.  Depending on your system,
+    instantiating a PurePath will return either a PurePosixPath or a
+    PureWindowsPath object.  You can also instantiate either of these classes
+    directly, regardless of your system.
+
+    On Posix, calling ``bytes()`` on a path gives the raw filesystem path as
```
Comment (merwok) [inline Lib/pathlib.py]: What do you think of:
```suggestion
    On Posix, calling ``bytes(path)`` gives the raw filesystem path as
```

File: Lib/pathlib.py [hunk 6]
```diff
@@ -707,6 +665,62 @@ def match(self, path_pattern, *, case_sensitive=None):
...
+    a bytes object, as encoded by ``os.fsencode()``. On Windows, the unicode
+    form is the canonical representation of fileysstem paths, and so calling
+    ``bytes()`` on a path is not recommended.
```
Comment (merwok) [inline Lib/pathlib.py]: Because it’s important in Python to understand the difference between referencing and calling a function, and the syntax is uniform so you can have `somefunc(args1)(args2)` or `dispatch[key](params)`, it’s a personal pet peeve to see parens used to denote functions like in old C documentation.  `bytes()` is not a function, `bytes` is!  But I know our sphinx doc do that (display `the bytes() function`) but at least the source code is `func:bytes` and there is a setting to add parens or not 🙂 

So take or leave this suggestion!

```suggestion
    a bytes object, as encoded by os.fsencode. On Windows, the unicode
    form is the canonical representation of fileysstem paths, and so calling
    bytes on a path is not recommended.
```

File: Lib/pathlib.py (hunk 1 above)
Comment (merwok) [reply Lib/pathlib.py:739]: I was wrong!  I think I had pydoc in mind, but it does show magic methods in the class doc and their docstrings.
So you could decide to remove the note about bytes from the class docstring, it’s not important enough to be there.

Maybe I mixed up pydoc with the docstring convention proposed PEP, or another tool.
I have a memory of a rule that doc for the constructor should be in the class docstring.  (Maybe because of the existence of \_\_new\_\_ and \_\_init\_\_?)

File: Lib/pathlib.py (hunk 6 above)
Comment (barneygale) [reply Lib/pathlib.py]: Interesting, makes sense - I'll bear that in mind for future! But I'll resolve this thread as we're deleting the paragraph :)

File: Lib/pathlib.py [hunk 7]
```diff
@@ -233,14 +233,10 @@ def __repr__(self):
         return "<{}.parents>".format(type(self._path).__name__)
 
 
-class PurePath(object):
-    """Base class for manipulating paths without I/O.
+class _BasePurePath(object):
```
Comment (AlexWaygood) [inline Lib/pathlib.py]: If we're touching this line anyway...

```suggestion
class _BasePurePath:
```

File: Lib/pathlib.py (hunk 3 above)
Comment (AlexWaygood) [reply Lib/pathlib.py]: I feel like a bigger issue here is... `os.PathLike` probably _should_ define `__slots__`? All the ABCs in `collections.abc` deliberately define `__slots__` for this very reason: so that other classes can subclass those ABCs and still be able to use the `__slots__` machinery to block the creation of instance dictionaries.

https://github.com/python/cpython/blob/c90a862cdcf55dc1753c6466e5fa4a467a13ae24/Lib/_collections_abc.py#L116-L118

https://github.com/python/cpython/blob/c90a862cdcf55dc1753c6466e5fa4a467a13ae24/Lib/_collections_abc.py#L131-L133

https://github.com/python/cpython/blob/c90a862cdcf55dc1753c6466e5fa4a467a13ae24/Lib/_collections_abc.py#L148-L150

https://github.com/python/cpython/blob/c90a862cdcf55dc1753c6466e5fa4a467a13ae24/Lib/_collections_abc.py#L192-L194

https://github.com/python/cpython/blob/c90a862cdcf55dc1753c6466e5fa4a467a13ae24/Lib/_collections_abc.py#L209-L211

(etc. etc.)

Comment (barneygale) [reply Lib/pathlib.py]: I can't find any discussion of it in [PEP-519](https://peps.python.org/pep-0519/), so perhaps it's just an oversight. @brettcannon penny for your thoughts?

File: Lib/os.py [hunk 8]
```diff
@@ -1079,6 +1079,8 @@ class PathLike(abc.ABC):
 
     """Abstract base class for implementing the file system path protocol."""
 
+    __slots__ = ()
```
Comment (AlexWaygood) [inline Lib/os.py]: Let's make this change in a separate PR/issue, I think

Comment (barneygale) [reply Lib/os.py]: Oh, OK!

Comment (AlexWaygood) [reply Lib/os.py]: Sorry if my previous comment was confusing!

I feel like the `.register()` thing works fine in pathlib for now, that this is a separate bugfix that should be considered separately (after we've heard Brett's thoughts), and that the git history will be easier to read if it's its own PR

Comment (barneygale) [reply Lib/os.py]: No worries, I was being silly. Issue and PR:

- #104898
- #104899

File: Lib/pathlib.py (hunk 3 above)
Comment (barneygale) [reply Lib/pathlib.py]: Fixed in #104898!

Comment (brettcannon) [reply Lib/pathlib.py]: @barneygale I take it you don't need that thought any more? (And FYI I live in Canada, so pennies are worthless to me, so you can keep your penny regardless. 😉)

File: Lib/test/test_pathlib.py [hunk 9]
```diff
@@ -37,7 +37,7 @@ def with_segments(self, *pathsegments):
         return type(self)(*pathsegments, session_id=self.session_id)
 
 
-class _BasePurePathTest(object):
+class _BaseBasePurePathTest(object):
```
Comment (barneygale) [inline Lib/test/test_pathlib.py]: This naming is very silly. It would be good to land #104829 first, which makes the test class hierarchy roughly match the pathlib class hierarchy as it is today, rather than as it was in the very early days of pathlib! This PR would change the class name here from `PurePathTest` to `BasePurePathTest`, which is much more reasonable.

File: Lib/pathlib.py (hunk 3 above)
Comment (barneygale) [reply Lib/pathlib.py]: Turns out the comment was right all along. If we subclass `os.PathLike`, we incur the cost of the ABC `__instancecheck__`. Previously this occurred in `_parse_args()`, which is the function that was referred to in the comment before I changed it to `__slots__`. D'oh. 

PR that reverts the pathlib change: #105073

**Comment by barneygale** [comment]
Would anyone be willing to review #104829? It blocks this PR. (please and thank you!)

**Comment by barneygale** [comment]
#104829 has landed, so test naming should be much improved! Think this is ready for review again.

**Comment by barneygale** [comment]
I've revised the implementation such that the methods are actually removed from `PurePath`. This will allow our future `AbstractPath` class to inherit `PurePath`, as it should conceptually:

image

(Contrast this with the [earlier proposed hierarchy](https://discuss.python.org/t/make-pathlib-extensible/3428/109?u=barneygale) where `PurePath` and `AbstractPath` are cousins.

The methods are moved to a new `_PurePathExt` class, which is inherited by `PurePosixPath`, `PureWindowsPath` and `Path`.

This change _will not_ be visible to users who try to instantiate `PurePath`, because they'll always get a `PurePosixPath` or `PureWindowsPath` instance back.

It _will_ be visible to users who subclass `PurePath`, but that has only just become possible in 3.12. As such I think it's worth backporting this into 3.12 for consistency.

**Comment by barneygale** [comment]
Here's a version of the diagram that includes the private `_PurePathExt` class.

image

**Comment by merwok** [comment]
Just a note that I’m meditating on https://hynek.me/articles/python-subclassing-redux/ and linked articles at the moment and wondering if controlling subclass explosion in pathlib should be a guideline 🙂

**Comment by barneygale** [comment]
Oh I totally agree. I'd prefer to add _only_ `pathlib.AbstractPath`, but I'm prevented by the fact that `PurePath` has an `__fspath__()` method :)

**Comment by barneygale** [comment]
I've converted this back to a draft as I'm still in two minds about how to implement this. It's possible I'll add `_AbstractPath` into this PR, or close this PR and log another, or something else. Sorry for the noise

**Comment by barneygale** [comment]
Friendship ended with #104810, now #106043 is my best friend.

**Comment by CAM-Gerlach** [comment]
@barneygale 

image
